modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java

```diff
@@ -196,10 +198,42 @@ public static BigDecimal toBigDecimal(Object o, int precision, int scale) {
 
         if (o instanceof Boolean) {
             throw new UnsupportedOperationException();
+        } else if (o instanceof Number) {
+            return toBigDecimal((Number) o, precision, scale);
+        } else {
+            return toBigDecimal(o.toString(), precision, scale);
+        }
+    }
+
+    /**
+     * Converts the given {@code BigDecimal} to a decimal with the given {@code precision} and {@code scale}
+     * according to SQL spec for CAST specification: General Rules, 8.
+     */
+    public static BigDecimal convertDecimal(BigDecimal value, int precision, int scale) {
+        assert precision > 0 : "Invalid precision: " + precision;
+
+        int defaultPrecision = IgniteTypeSystem.INSTANCE.getDefaultPrecision(SqlTypeName.DECIMAL);
+        if (precision == defaultPrecision) {
+            // This branch covers at least one known case: access to dynamic parameter from context.
+            // In this scenario precision = DefaultTypePrecision, because types for dynamic params
+            // are created by toSql(createType(param.class)).
+            return value;
+        }
+
+        boolean nonZero = !value.unscaledValue().equals(BigInteger.ZERO);
+
+        if (nonZero && scale > precision) {
+            throw new SqlException(QUERY_INVALID_ERR, "Numeric overflow");
```

Comment:
why this raises "Numeric overflow" ? i suppose this is something like: erroneous format or smth else ?

Reply:
@zstan I looked at PostgeSQL and it returns the same error in both cases.